[#2] 회원가입 구현

seohoon-dev opened the pull request:
#2

seohoon-dev committed:
회원가입

f-lab-dev commented on src/main/java/com/hoon/foodrocket/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http
+                .cors().disable()
```
여기서 설정 하나하나의 역할을 주석으로 달아주시면 좋을 것 같습니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/EmailExistedException.java:
```diff
@@ -0,0 +1,7 @@
+package com.hoon.foodrocket.application;
+
+public class EmailExistedException extends RuntimeException {
```
좀 더 자세한 예외를 상속받아도 좋을 것 같습니다. `IlegalStateException`도 괜찮아보이네요~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/UserService.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Optional;
+
+@Service
+@Transactional
```
클래스단위로 Transactional을 달아준 이유가 있을까요?

seohoon-dev replied:
클래스 단위로 Transactional 을 붙여주면 클래스 내부에 있는 메소드는 개별적으로 Transactional 붙여준 것과 동일한 기능을 한다고하여 그렇게 했습니다.
다만 클래스와 메소드에 각각 Transactional 을 붙여줄 경우에는 메소드에 적용된 Transactional 이 우선순위를 갖는다고 봤습니다.

f-lab-dev replied:
그럼 Transactional이 필요없는 메소드를 추가했는데 자신도 모르게 이 Transactional이 적용되지는 않을까요?

seohoon-dev replied:
그 부분까지 생각하지 못했네요. 확실히 불필요하게 Transactional이 적용되는 경우가 있을 것 같습니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/domain/User.java:
```diff
@@ -0,0 +1,32 @@
+@Builder
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
```
여기서도 각 어노테이션의 역할을 주석으로 달아주면 좋을 것 같습니다.

f-lab-dev commented on src/main/java/com/hoon/foodrocket/interfaces/UserController.java:
```diff
@@ -0,0 +1,32 @@
+
+        User user = userService.registerUser(email, name, password);
+
+        String url = "/users/" +user.getId();
```
URL을 굳이 써줄필요는 없습니다. URL을 보내고싶으시다면 보내야하는 이유도 같이 주석으로 적어두면 좋을 것 같네요

f-lab-dev commented on src/main/java/com/hoon/foodrocket/domain/UserRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.Optional;
+
+public interface UserRepository extends CrudRepository<User, Long> {
```
JPA를 쓰는것도 좋지만 SQL을 잘 알아야하니 MySQL 쿼리작성법도 공부해두시면 좋을 것 같네요

seohoon-dev committed:
리뷰 내용 반영

- 주석 추가
- 예외 상속 변경(RuntimeException -> IllegalStateException)
- ResponseEntity URL 제거

f-lab-dev commented on src/main/java/com/hoon/foodrocket/SecurityJavaConfig.java:
```diff
@@ -15,10 +15,10 @@
-                .csrf().disable()
-                .formLogin().disable()
-                .headers().frameOptions().disable();
+                .cors().disable()   // 외부서버 데이터를 받도록 cors 옵션 비활성화
```
CORS 관련 설정을 하셨으니 CORS가 어떤 원리로 이루어지는지, 이 옵션을 disable하면 웹에서는 어떤일이 일어나는지 알아두면 좋을 것 같습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/SecurityJavaConfig.java:
```diff
@@ -15,10 +15,10 @@
-                .formLogin().disable()
-                .headers().frameOptions().disable();
+                .cors().disable()   // 외부서버 데이터를 받도록 cors 옵션 비활성화
+                .csrf().disable()   // csrf 프로텍션 비활성화
```
CSRF도 마찬가지이고요

seohoon-dev committed:
회원가입 구현

- mybatis 적용

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/UserService.java:
```diff
@@ -1,31 +1,40 @@
 
 @Service
-@Transactional
+@Transactional  // 메소드나 클래스에 트랜잭션 속성을 설명
```
단순히 설명한다라기보다는 이 어노테이션이 달림으로써 백그라운드에서 어떤일이 일어나고, 이 어노테이션이 달리면 어떤 효과가 적용되는지를 설명하면 좋을 것 같습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/application/EmailNotExistedException.java:
```diff
@@ -0,0 +1,7 @@
+package com.hoon.foodrocket.application;
+
+public class EmailNotExistedException extends IllegalStateException {
```
[팁] 모든 예외상황의 클래스를 다 만들어줄 필요는 없습니다. 왠만해서는 `new IllegalStateException("Email is not exist")` 이정도로 처리할 수 있기 때문입니다. 다만 구체적인 예외상황에 따라 다르게 처리해줄 필요가 있을 때에는 이렇게 정의해주는 것이 좋습니다. (혹시 이해가 안돼시면 다시 코멘트 주세요)

seohoon-dev replied:
구체적인 예외상황에 대한 예시나 문서가 있다면 이해에 도움이 될 것 같습니다.

f-lab-dev replied:
"로그인할 때 `EmailNotExistedException`의 경우에는 회원가입 페이지로 보내준다." 이런 요구사항이 있다면 이렇게 따로 `EmailNotExistedException`을 정의하면 더 간단하게 문제를 해결할 수 있습니다. 그런 상황을 말씀드린 것이였습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/UserController.java:
```diff
@@ -1,28 +1,32 @@
     private UserService userService;
 
-    @PostMapping("/users")
+    @GetMapping("/user")
```
URL은 보통 복수형으로 표현합니다. `/users/{userId}` 이런식으로요
아래의 문서가 도움이 될 것 같네요
https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/UserController.java:
```diff
@@ -1,28 +1,32 @@
 
-    @PostMapping("/users")
+    @GetMapping("/user")
+    public User detail(@RequestParam("email") String email) {
```
보통 유저를 조회할때는 userId로 조회합니다. email을 받은 의도가 따로 있으실까요?

seohoon-dev replied:
회원가입시 아이디 대신 이메일을 받고 있습니다.
별도의 아이디를 받는게 좋을까요?

f-lab-dev replied:
userId라고 한 것은 DB에 insert할 때 자동으로 발급되는 ID값을 말한 것이였습니다~

f-lab-dev commented on src/main/java/com/hoon/foodrocket/controller/UserController.java:
```diff
@@ -1,28 +1,32 @@
-        User user = userService.registerUser(email, name, password);
+        userService.registerUser(email, name, password);
 
         return ResponseEntity.status(HttpStatus.CREATED).build();
```
[팁] `ResponseEntity.ok()`도 있습니다

seohoon-dev replied:
계정 생성에 대한 부분이라 created를 사용했는데 ok가 좀 더 괜찮은 표현일까요?

f-lab-dev replied:
아 이 부분은 제가 잘못봤네요 CREATED를 잘못봤습니다

seohoon-dev committed:
리뷰 내용 반영

- Transactional 어노테이션 설명 추가
- 간단한 예외상황을 표현한 클래스 제거
- 컨트롤러 url 복수형으로 표현
- ReponseEntity status 옵션 ok로 변경

seohoon-dev committed:
리뷰 내용 반영

seohoon-dev committed:
mysql 적용
